apps/web/src/app/(profile)/[username]/_components/dashboard/_actions.ts

```diff
@@ -167,72 +185,69 @@ export async function getSolvedChallenges(userId: string) {
   };
 }
 
-export interface BadgeInfo {
-  // eslint-disable-next-line @typescript-eslint/sort-type-constituents
-  slug: 'aot-2023-bronze' | 'aot-2023-silver' | 'aot-2023-gold' | 'aot-2023-platinum';
+export interface Badges<T> {
+  slug: T;
   name: string;
+  shortName: string;
 }
 
-export async function getBadges(userId: string): Promise<BadgeInfo[]> {
-  const badges: BadgeInfo[] = [];
+type BadgeTypes = AotBadges | DifficultyBadges | SolutionBadges;
+export type BadgeObj<T> = {
+  [key in BadgeTypes]?: {
+    slug: key;
+    name: string;
+    shortName: string;
+  };
+};
 
-  const holidayTrack = await prisma.track.findFirst({
-    where: {
-      slug: 'advent-of-typescript-2023',
-    },
-    include: {
-      trackChallenges: {
-        orderBy: {
-          orderId: 'asc',
-        },
-        include: {
-          challenge: {
-            include: {
-              submission: {
-                where: {
-                  userId,
-                },
-              },
-            },
-          },
-        },
-      },
-      enrolledUsers: {
-        where: {
-          id: userId,
-        },
-        select: {
-          id: true,
-        },
-      },
-    },
-  });
+export type AllBadges = Badges<BadgeTypes>;
+export type AllBadgeObjs = BadgeObj<BadgeTypes>;
 
-  const numberOfAttemptedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return (trackChallenge.challenge.submission?.length ?? 0) > 0;
-    }).length ?? 0;
+export type BadgeFn = ({ userId, badges }: { userId: string; badges: AllBadgeObjs }) => Promise<AllBadgeObjs>;
 
-  if (numberOfAttemptedHolidayChallenges > 0) {
-    badges.push({ slug: 'aot-2023-bronze', name: 'Advent of TypeScript 2023 Bronze' });
-  }
+const badgeCalculations: BadgeFn[] = [
+  adventBadgesFn,
+  difficultyBadgesFn,
+  sharedSolutionsBadgesFn,
+];
 
-  const numberOfCompletedHolidayChallenges =
-    holidayTrack?.trackChallenges.filter((trackChallenge) => {
-      return trackChallenge.challenge.submission?.some((submission) => submission.isSuccessful);
-    }).length ?? 0;
+// TODO: If the submission is successful this gets called, is the right spot?
+export async function fillInMissingBadges(userId: string): Promise<AllBadgeObjs> {
```

Comment:
This endpoint is called when a submission is successful to recalculate and fill in any missing badges.